### PR TITLE
Add secrets CRUD API 

### PR DESCRIPTION
**Pull request by stefanprodan**
- implement HTTP GET/POST/PUT/DEL for secret type
- omit the secret value from GET result
- return only error codes and log the error body
- use label `app.kubernetes.io/managed-by: openfaas` to filter secrets created by the operator
- add happy path CRUD tests
- update Kubernetes deps to v1.12 (fix for klog)
- update faas packages to get the secrets handler and struct

Fix: #69 

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?
Tested on GKE with curl:

Create secret:

```bash
curl -vd '{"name":"test","value":"test"}' -X POST http://localhost:8081/system/secrets

 HTTP/1.1 202 Accepted
```

List secrets:

```bash
curl -s http://localhost:8081/system/secrets

[{"name":"test"},{"name":"test2"}]
```

Update secret:

```bash
curl -vd '{"name":"test","value":"test update"}' -X PUT http://localhost:8081/system/secrets

HTTP/1.1 202 Accepted
```

Delete secret:

```bash
curl -vd '{"name":"test"}' -X DELETE http://localhost:8081/system/secrets

HTTP/1.1 202 Accepted
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Impact to existing users

None

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
@stefanprodan please could you provide your curl statements and results? It'll be useful for giving to the others who are doing faas-swarm etc whilst we have no CLI command available.

**Comment by stefanprodan** [comment]
@alexellis @LucasRoesler I've updated the PR description with the curl tests examples and results.

Make sure you apply the RBAC before testing since the operator now needs full access to secrets, before was readonly.

File: pkg/server/secret.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+			glog.Infof("Secret %s deleted", secret.GetName())
+			w.WriteHeader(http.StatusAccepted)
+		default:
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+	}
+}
+
+func getSecret(namespace string, r io.Reader) (*corev1.Secret, error) {
```
Comment (LucasRoesler) [inline pkg/server/secret.go]: Everything looks good and my manual tests worked. The only thing I thought was confusing in the code was the name of this method.  I think `parseSecret` would be more accurate . Other than that 💯

File: pkg/server/secret.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	if err != nil {
+		return nil, err
+	}
+	secret := &corev1.Secret{
+		Type: corev1.SecretTypeOpaque,
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: namespace,
+			Labels: map[string]string{
+				"owner": "openfaas",
```
Comment (LucasRoesler) [inline pkg/server/secret.go]: I was thinking more about this label, I think k8s is recommending specificy label formats now https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels

I wonder if `app.kubernetes.io/part-of` or `app.kubernetes.io/managed-by` would be a better label name?

Comment (stefanprodan) [reply pkg/server/secret.go]: Yes you're right, I totally forgot about that convention. I'll change it.

File: README.md [hunk 3]
```diff
@@ -105,13 +105,13 @@ Start OpenFaaS controller (assumes you have a working kubeconfig on the machine)
 
 ```bash
 $ go build \
-  && ./openfaas-operator -kubeconfig=$HOME/.kube/config -logtostderr=true -v=4
```
Comment (alexellis) [inline README.md]: Does that also need removing from helm, or is it a no-harm change?

File: artifacts/operator-crd.yaml [hunk 4]
```diff
@@ -45,18 +45,3 @@ spec:
                 memory:
                   type: string
                   pattern: "^[0-9]+(Mi|Gi)"
-            secrets:
```
Comment (alexellis) [inline artifacts/operator-crd.yaml]: Why did this get removed?

File: main.go [hunk 5]
```diff
@@ -41,6 +41,18 @@ func main() {
 	flag.Set("logtostderr", "true")
 	flag.Parse()
 
+	klogFlags := flag.NewFlagSet("klog", flag.ExitOnError)
```
Comment (alexellis) [inline main.go]: Please extract to a function with a name to document what's going on or a comment.

File: pkg/server/secret.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+)
+
+// makeSecretHandler provides the secrets CRUD endpoint
+func makeSecretHandler(namespace string, kube kubernetes.Interface) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.Body != nil {
+			defer r.Body.Close()
+		}
+
+		switch r.Method {
```
Comment (alexellis) [inline pkg/server/secret.go]: Please can you split each out into its own method for maintainability. I've asked @bartsmykla to do the same for faas-swarm.

File: pkg/server/secret.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+		case http.MethodGet:
+			selector := fmt.Sprintf("%s=%s", secretLabel, secretLabelValue)
+			res, err := kube.CoreV1().Secrets(namespace).List(metav1.ListOptions{LabelSelector: selector})
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
```
Comment (alexellis) [inline pkg/server/secret.go]: In faas-netes we write the error back to the client. I think we should do that here too, or if you're worried about leaking data then at last write what happened without quoting the error.

File: pkg/server/secret.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+			w.WriteHeader(http.StatusAccepted)
+		case http.MethodPut:
+			newSecret, err := parseSecret(namespace, r.Body)
+			if err != nil {
+				w.WriteHeader(http.StatusBadRequest)
+				glog.Errorf("Secret unmarshal error: %v", err)
+				return
+			}
+			secret, err := kube.CoreV1().Secrets(namespace).Get(newSecret.GetName(), metav1.GetOptions{})
+			if errors.IsNotFound(err) {
```
Comment (alexellis) [inline pkg/server/secret.go]: I think that we should write a text error back in the body.

File: pkg/server/secret.go [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+	}
+}
+
+func parseSecret(namespace string, r io.Reader) (*corev1.Secret, error) {
+	body, _ := ioutil.ReadAll(r)
+	req := requests.Secret{}
+	err := json.Unmarshal(body, &req)
```
Comment (alexellis) [inline pkg/server/secret.go]: Can this be inlined whilst making the other edits?

File: README.md [hunk 10]
```diff
@@ -145,6 +145,8 @@ kubectl get all
 
 ### API calls
 
+#### Function CRUD
```
Comment (alexellis) [inline README.md]: Replace CRUD with management?

File: artifacts/operator-crd.yaml (hunk 4 above)
Comment (stefanprodan) [reply artifacts/operator-crd.yaml]: It doesn't work if there are no secrets specified, I've synced the CRD that's in the faas-netes chart and that one is working fine.

File: README.md (hunk 3 above)
Comment (stefanprodan) [reply README.md]: `logtostderr=true` is now the default so no need to specify it, there is a no-harm change

File: main.go (hunk 5 above)
Comment (stefanprodan) [reply main.go]: the comment is 2 lines below, it syncs the the glog flags with klog

File: pkg/server/secret.go (hunk 7 above)
Comment (stefanprodan) [reply pkg/server/secret.go]: None of the API handlers are returning a body on error, just status codes. If we want this then we should define an error struct and refactor the whole API.

Comment (alexellis) [reply pkg/server/secret.go]: We do write the error to the client, here's an example: https://github.com/openfaas/faas-netes/blob/master/handlers/deploy.go#L79

No need to refactor to an error struct, just write the error and a sensible message to the client like above?

File: main.go (hunk 5 above)
Comment (alexellis) [reply main.go]: Can you extract to a function please?

File: artifacts/operator-crd.yaml (hunk 4 above)
Comment (alexellis) [reply artifacts/operator-crd.yaml]: If this is fixing a bug can we track that separately or in the issue tracker?

Comment (stefanprodan) [reply artifacts/operator-crd.yaml]: this was fixed in faas-netes 4 months ago https://github.com/openfaas/faas-netes/commit/da1ad5abd53b9628cbbebec1f70b2b4f27ff1616

**Comment by stefanprodan** [comment]
@alexellis the latest commit addresses your comments.

**Comment by alexellis** [comment]
I would still like to see the handler split out into multiple functions per action / HTTP method.

I'll merge and this can be done afterwards. cc @LucasRoesler 

Alex

**Pull request merged**